File: android/app/src/main/java/com/openrung/vpn/PunchRecoveryCircuitBreaker.kt
package com.openrung.vpn

import kotlinx.coroutines.delay
import kotlin.random.Random

internal sealed interface PunchRecoveryDecision {
    val delayMs: Long
    val rapidFailureCount: Int
    val directUptimeMs: Long

    data class RetryDirect(
        override val delayMs: Long,
        override val rapidFailureCount: Int,
        override val directUptimeMs: Long,
        val countedFailure: Boolean,
    ) : PunchRecoveryDecision

    data class UseRelayHub(
        override val delayMs: Long,
        override val rapidFailureCount: Int,
        override val directUptimeMs: Long,
    ) : PunchRecoveryDecision
}

/**
 * Bounds recovery churn from a direct path that repeatedly reaches CONNECTED and then dies.
 *
 * State is per relay so one unstable volunteer does not disable direct paths to other volunteers.
 * The service owns this object on its Main dispatcher; no internal synchronization is needed.
 */
internal class PunchRecoveryCircuitBreaker(
    private val stablePathMs: Long = STABLE_PATH_MS,
    private val maxRapidFailures: Int = MAX_RAPID_FAILURES,
    private val initialBackoffMs: Long = INITIAL_BACKOFF_MS,
    private val maxBackoffMs: Long = MAX_BACKOFF_MS,
    private val chooseJitteredDelay: (minimumMs: Long, maximumMs: Long) -> Long =
        { minimumMs, maximumMs ->
            if (minimumMs == maximumMs) {
                minimumMs
            } else {
                Random.nextLong(minimumMs, maximumMs + 1)
            }
        },
) {
    init {
        require(stablePathMs > 0) { "stablePathMs must be positive" }
        require(maxRapidFailures > 0) { "maxRapidFailures must be positive" }
        require(initialBackoffMs > 0) { "initialBackoffMs must be positive" }
        require(maxBackoffMs >= initialBackoffMs) {
            "maxBackoffMs must be at least initialBackoffMs"
        }
    }

    private data class RelayState(
        var rapidFailureCount: Int = 0,
        var directConnectedAtMs: Long? = null,
        var circuitOpen: Boolean = false,
    )

    private val relayStates = mutableMapOf<String, RelayState>()

    fun markDirectConnected(relayId: String, nowElapsedMs: Long) {
        require(relayId.isNotBlank()) { "relayId must not be blank" }
        relayStates.getOrPut(relayId, ::RelayState).directConnectedAtMs = nowElapsedMs
    }

    fun allowsDirectPunch(relayId: String): Boolean = relayStates[relayId]?.circuitOpen != true

    /**
     * Records a lost direct path and returns the recovery action. A physical-network outage passes
     * [countTowardBreaker] as false: it preserves any earlier rapid-failure streak, adds no delay,
     * and never opens the circuit. A path that survived [stablePathMs] clears the earlier streak
     * before the current loss is considered.
     */
    fun onDirectPathLost(
        relayId: String,
        nowElapsedMs: Long,
        countTowardBreaker: Boolean,
    ): PunchRecoveryDecision {
        require(relayId.isNotBlank()) { "relayId must not be blank" }
        val state = relayStates.getOrPut(relayId, ::RelayState)
        val connectedAtMs = state.directConnectedAtMs
        state.directConnectedAtMs = null
        val directUptimeMs = connectedAtMs
            ?.let { (nowElapsedMs - it).coerceAtLeast(0) }
            ?: 0

        if (directUptimeMs >= stablePathMs) {
            state.rapidFailureCount = 0
            state.circuitOpen = false
        }

        if (!countTowardBreaker) {
            return PunchRecoveryDecision.RetryDirect(
                delayMs = 0,
                rapidFailureCount = state.rapidFailureCount,
                directUptimeMs = directUptimeMs,
                countedFailure = false,
            )
        }

        state.rapidFailureCount++
        val delayMs = recoveryDelayMs(state.rapidFailureCount)
        if (state.rapidFailureCount >= maxRapidFailures) {
            state.circuitOpen = true
            return PunchRecoveryDecision.UseRelayHub(
                delayMs = delayMs,
                rapidFailureCount = state.rapidFailureCount,
                directUptimeMs = directUptimeMs,
            )
        }

        return PunchRecoveryDecision.RetryDirect(
            delayMs = delayMs,
            rapidFailureCount = state.rapidFailureCount,
            directUptimeMs = directUptimeMs,
            countedFailure = true,
        )
    }

    /** Explicit user connect/disconnect starts a new recovery epoch. */
    fun reset() {
        relayStates.clear()
    }

    private fun recoveryDelayMs(failureCount: Int): Long {
        var nominalMs = initialBackoffMs
        repeat((failureCount - 1).coerceAtLeast(0)) {
            nominalMs = if (nominalMs >= maxBackoffMs / 2) {
                maxBackoffMs
            } else {
                (nominalMs * 2).coerceAtMost(maxBackoffMs)
            }
        }
        val jitterMs = nominalMs / JITTER_DIVISOR
        val minimumMs = (nominalMs - jitterMs).coerceAtLeast(1)
        val maximumMs = (nominalMs + jitterMs).coerceAtMost(maxBackoffMs)
        return chooseJitteredDelay(minimumMs, maximumMs).coerceIn(minimumMs, maximumMs)
    }

    companion object {
        internal const val STABLE_PATH_MS = 5 * 60_000L
        internal const val MAX_RAPID_FAILURES = 3
        internal const val INITIAL_BACKOFF_MS = 2_000L
        internal const val MAX_BACKOFF_MS = 30_000L
        private const val JITTER_DIVISOR = 5L // +/-20 percent
    }
}

/** Kept as a suspend operation so disconnect or a manual reconnect cancels pending backoff. */
internal suspend fun PunchRecoveryDecision.awaitBackoff() {
    if (delayMs > 0) delay(delayMs)
}

File: ios/PacketTunnel/PunchRecoveryCircuitBreaker.swift
import Foundation

/// Recovery action after a previously connected direct punch path is lost.
///
/// Every case retains the same measurements so the caller can log recovery without consulting
/// mutable breaker state. `countedFailure` is false only when the loss was caused by a physical
/// network outage.
enum PunchRecoveryDecision: Equatable, Sendable {
    case retryDirect(
        delayMilliseconds: UInt64,
        rapidFailureCount: Int,
        directUptimeMilliseconds: UInt64,
        countedFailure: Bool
    )
    case useRelayHub(
        delayMilliseconds: UInt64,
        rapidFailureCount: Int,
        directUptimeMilliseconds: UInt64,
        countedFailure: Bool
    )

    var delayMilliseconds: UInt64 {
        switch self {
        case let .retryDirect(delay, _, _, _), let .useRelayHub(delay, _, _, _):
            return delay
        }
    }

    var rapidFailureCount: Int {
        switch self {
        case let .retryDirect(_, count, _, _), let .useRelayHub(_, count, _, _):
            return count
        }
    }

    var directUptimeMilliseconds: UInt64 {
        switch self {
        case let .retryDirect(_, _, uptime, _), let .useRelayHub(_, _, uptime, _):
            return uptime
        }
    }

    var countedFailure: Bool {
        switch self {
        case let .retryDirect(_, _, _, counted), let .useRelayHub(_, _, _, counted):
            return counted
        }
    }

    /// Keeps recovery backoff owned by the caller's task, so disconnecting or manually
    /// reconnecting cancels the pending retry.
    func awaitBackoff() async throws {
        guard delayMilliseconds > 0 else { return }
        let nanoseconds = delayMilliseconds.multipliedReportingOverflow(by: 1_000_000)
        try await Task.sleep(
            nanoseconds: nanoseconds.overflow ? UInt64.max : nanoseconds.partialValue
        )
    }
}

/// Bounds recovery churn from a direct path that repeatedly reaches connected and then dies.
///
/// State is isolated by relay. The owning packet-tunnel task must serialize access, matching the
/// Android service's single-dispatcher ownership; this type performs no internal synchronization.
final class PunchRecoveryCircuitBreaker {
    static let stablePathMilliseconds: UInt64 = 5 * 60_000
    static let maximumRapidFailures = 3
    static let initialBackoffMilliseconds: UInt64 = 2_000
    static let maximumBackoffMilliseconds: UInt64 = 30_000

    typealias JitterSelector = (_ allowedDelay: ClosedRange<UInt64>) -> UInt64

    private struct RelayState {
        var rapidFailureCount = 0
        var directConnectedAtMilliseconds: UInt64?
        var circuitOpen = false
    }

    private let stablePathMilliseconds: UInt64
    private let maximumRapidFailures: Int
    private let initialBackoffMilliseconds: UInt64
    private let maximumBackoffMilliseconds: UInt64
    private let chooseJitteredDelay: JitterSelector
    private var relayStates: [String: RelayState] = [:]

    init(
        stablePathMilliseconds: UInt64 = PunchRecoveryCircuitBreaker.stablePathMilliseconds,
        maximumRapidFailures: Int = PunchRecoveryCircuitBreaker.maximumRapidFailures,
        initialBackoffMilliseconds: UInt64 =
            PunchRecoveryCircuitBreaker.initialBackoffMilliseconds,
        maximumBackoffMilliseconds: UInt64 =
            PunchRecoveryCircuitBreaker.maximumBackoffMilliseconds,
        chooseJitteredDelay: @escaping JitterSelector = { UInt64.random(in: $0) }
    ) {
        precondition(stablePathMilliseconds > 0, "stablePathMilliseconds must be positive")
        precondition(maximumRapidFailures > 0, "maximumRapidFailures must be positive")
        precondition(initialBackoffMilliseconds > 0, "initialBackoffMilliseconds must be positive")
        precondition(
            maximumBackoffMilliseconds >= initialBackoffMilliseconds,
            "maximumBackoffMilliseconds must be at least initialBackoffMilliseconds"
        )
        self.stablePathMilliseconds = stablePathMilliseconds
        self.maximumRapidFailures = maximumRapidFailures
        self.initialBackoffMilliseconds = initialBackoffMilliseconds
        self.maximumBackoffMilliseconds = maximumBackoffMilliseconds
        self.chooseJitteredDelay = chooseJitteredDelay
    }

    func markDirectConnected(relayID: String, nowElapsedMilliseconds: UInt64) {
        requireRelayID(relayID)
        var state = relayStates[relayID] ?? RelayState()
        state.directConnectedAtMilliseconds = nowElapsedMilliseconds
        relayStates[relayID] = state
    }

    func allowsDirectPunch(relayID: String) -> Bool {
        relayStates[relayID]?.circuitOpen != true
    }

    /// Records a lost direct path and returns the recovery action.
    ///
    /// Passing `countTowardBreaker` as false for a physical-network outage preserves any earlier
    /// rapid-failure streak, adds no delay, and never opens the circuit. A path that survived the
    /// stable-path threshold clears its earlier streak before the current loss is considered.
    func onDirectPathLost(
        relayID: String,
        nowElapsedMilliseconds: UInt64,
        countTowardBreaker: Bool
    ) -> PunchRecoveryDecision {
        requireRelayID(relayID)
        var state = relayStates[relayID] ?? RelayState()
        let connectedAt = state.directConnectedAtMilliseconds
        state.directConnectedAtMilliseconds = nil
        let directUptime: UInt64
        if let connectedAt, nowElapsedMilliseconds >= connectedAt {
            directUptime = nowElapsedMilliseconds - connectedAt
        } else {
            directUptime = 0
        }

        if directUptime >= stablePathMilliseconds {
            state.rapidFailureCount = 0
            state.circuitOpen = false
        }

        guard countTowardBreaker else {
            relayStates[relayID] = state
            return .retryDirect(
                delayMilliseconds: 0,
                rapidFailureCount: state.rapidFailureCount,
                directUptimeMilliseconds: directUptime,
                countedFailure: false
            )
        }

        state.rapidFailureCount += 1
        let delay = recoveryDelayMilliseconds(failureCount: state.rapidFailureCount)
        if state.rapidFailureCount >= maximumRapidFailures {
            state.circuitOpen = true
            relayStates[relayID] = state
            return .useRelayHub(
                delayMilliseconds: delay,
                rapidFailureCount: state.rapidFailureCount,
                directUptimeMilliseconds: directUptime,
                countedFailure: true
            )
        }

        relayStates[relayID] = state
        return .retryDirect(
            delayMilliseconds: delay,
            rapidFailureCount: state.rapidFailureCount,
            directUptimeMilliseconds: directUptime,
            countedFailure: true
        )
    }

    /// Explicit user connect or disconnect starts a new recovery epoch.
    func reset() {
        relayStates.removeAll()
    }

    private func recoveryDelayMilliseconds(failureCount: Int) -> UInt64 {
        var nominal = initialBackoffMilliseconds
        if failureCount > 1 {
            for _ in 1..<failureCount {
                if nominal >= maximumBackoffMilliseconds / 2 {
                    nominal = maximumBackoffMilliseconds
                } else {
                    nominal = min(nominal * 2, maximumBackoffMilliseconds)
                }
            }
        }

        let jitter = nominal / 5
        let minimum = max(nominal - jitter, 1)
        let addition = nominal.addingReportingOverflow(jitter)
        let maximum = min(
            addition.overflow ? UInt64.max : addition.partialValue,
            maximumBackoffMilliseconds
        )
        let allowedDelay = minimum...maximum
        return min(max(chooseJitteredDelay(allowedDelay), minimum), maximum)
    }

    private func requireRelayID(_ relayID: String) {
        precondition(
            relayID.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty == false,
            "relayID must not be blank"
        )
    }
}
